pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,219 @@
+		}
+	}
+}
+
+// delete removes the provided IP-to-security-identity mapping from both caches
+// within ipc.
+func (ipc *IPCache) delete(endpointIP string) {
+	ipc.mutex.Lock()
+	defer ipc.mutex.Unlock()
+	ipc.deleteLocked(endpointIP)
+}
+
+// LookupByIP returns the corresponding security identity that endpoint IP maps
+// to within the provided IPCache, as well as if the corresponding entry exists
+// in the IPCache.
+func (ipc *IPCache) LookupByIP(endpointIP string) (identity.NumericIdentity, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	identity, exists := ipc.ipToIdentityCache[endpointIP]
+	return identity, exists
+}
+
+// LookupByIdentity returns the set of endpoint IPs that have security identity
+// ID, as well as if the corresponding entry exists in the IPCache.
+func (ipc *IPCache) LookupByIdentity(id identity.NumericIdentity) (map[string]struct{}, bool) {
+	ipc.mutex.RLock()
+	defer ipc.mutex.RUnlock()
+	ips, exists := ipc.identityToIPCache[id]
+	return ips, exists
+}
+
+// IPIdentityMappingOwner is the interface the owner of an identity allocator
+// must implement
+type IPIdentityMappingOwner interface {
+	// OnIPIdentityCacheChange will be called whenever there the state of the
+	// IPCache has changed.
+	OnIPIdentityCacheChange()
+}
+
+// GetIPIdentityMapModel returns all known endpoint IP to security identity mappings
+// stored in the key-value store.
+func GetIPIdentityMapModel() {
+	// TODO (ianvernon) return model of ip to identity mapping. For use in CLI.
+	// see GH-2555
+}
+
+func ipIdentityWatcher(owner IPIdentityMappingOwner) {
+
+	for {
+		watcher := kvstore.ListAndWatch("endpointIPWatcher", IPIdentitiesPath, 512)
+
+		// Get events from channel as they come in.
+		for event := range watcher.Events {
+
+			var (
+				cacheChanged bool
+				ipIDPair     identity.IPIdentityPair
+			)
+
+			err := json.Unmarshal(event.Value, &ipIDPair)
+			if err != nil {
+				log.Errorf("not adding entry to ip cache; error unmarshaling identity from key-value store: %s", err)
+				continue
+			}
+
+			// Synchronize local caching of endpoint IP to ipIDPair mapping with
+			// operation key-value store has informed us about.
+
+			ipStr := ipIDPair.IP.String()
+
+			cachedIdentity, exists := IPIdentityCache.LookupByIP(ipStr)
+			endpointIPs, _ := IPIdentityCache.LookupByIdentity(ipIDPair.ID)
+
+			switch event.Typ {
+			case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+				if !exists || cachedIdentity != ipIDPair.ID {
+					IPIdentityCache.upsert(ipStr, ipIDPair.ID)
+					cacheChanged = true
+
+					// Update XDS Cache as well.
+					ipStrings := make([]string, 0, len(endpointIPs))
+					for endpointIP := range endpointIPs {
+						ipStrings = append(ipStrings, endpointIP)
+					}
+					sort.Strings(ipStrings)
+					envoy.NetworkPolicyHostsCache.Upsert(envoy.NetworkPolicyHostsTypeURL, ipIDPair.ID.StringID(), &envoyAPI.NetworkPolicyHosts{Policy: uint64(ipIDPair.ID), HostAddresses: ipStrings}, false)
```

Comment:
The envoy package dependency is quite ugly, let's get rid of it after the initial merge.

Reply:
What would the best way be to get rid of it? I'm interested in how such a design is achieved so I can have cleaner code in the future.

Reply:
The host cache needs to be factored out of the envoy package, the envoy package should define an interface that the host cache implements. That way, your package can reuse the code and share the cache directly for XDS use but does not depend on anything that is envoy specific.

Reply:
Created https://github.com/cilium/cilium/issues/3087 to track this.

Reply:
An alternative is to define a callback type in this package, and have others register conforming callbacks. This has the advantage that the dependency direction reflects the flow of data (in that the envoy package knows about ipcache and is using it, so it imports it), instead of a push scheme like we have now. It also allows this attachment to happen in a third package, allowing generic types to be wired together without the type system needing to know.
If we intended to have a global host cache then what I said is less useful, but it still retains the niceties around reflecting which direction data is flowing.